File: 2024/02_red-nosed_reports/py/solution.py
with open("02_input.txt") as file:
    data = []
    for line in file:
        data.append([int(x) for x in line.strip().split()])

def is_safe(diffs):
    if all(x > 0 for x in diffs) or all(x < 0 for x in diffs):
        MIN = min(abs(x) for x in diffs)
        MAX = max(abs(x) for x in diffs)

        result = MIN > 0 and MAX <= 3
        return result

    return False

def is_really_safe(levels):

    if is_safe(get_diffs(levels)):
        return True

    for i in range(len(levels)):
        new_levels = levels[:i] + levels[i + 1:]
        if is_safe(get_diffs(new_levels)):
            return True

    return False

def get_diffs(levels):
    diffs = []
    for i in range(len(levels) - 1):
        diffs.append(levels[i] - levels[i + 1])
    return diffs


def part1():
    return sum(1 for levels in data if is_safe(get_diffs(levels)))

def part2():
    return sum(1 for levels in data if is_really_safe(levels))

for levels in data:
    print(levels, is_really_safe(levels))

print(f"Part 1: {part1()}")
print(f"Part 2: {part2()}")

File: 2024/02_red-nosed_reports/rs/solution.rs
use std::fs;
use std::io;

fn read_data(file_path: &str) -> io::Result<Vec<Vec<i32>>>
{
    let contents = fs::read_to_string(file_path)?;
    let mut result = Vec::new();

    for line in contents.lines()
    {
        let mut list = Vec::new();
        let parts : Vec<&str> = line.split_whitespace().collect();
        for part in parts
        {
            list.push(part.parse::<i32>().unwrap());
        }

        result.push(list);
    }

    Ok(result)
}

fn is_valid(list: &Vec<i32>) -> bool
{
    let mut correct = true;
    let decreasing = list[0] > list[1];

    for window in list.windows(2)
    {
        let curr_decrease = window[0] > window[1];
        if curr_decrease != decreasing
        {
            correct = false;
            break;
        }

        let diff = (window[0] - window[1]).abs();
        if diff == 0 || diff > 3
        {
            correct = false;
            break;
        }
    }

    correct
}

fn part1(data: &Vec<Vec<i32>>) -> i32
{
    let mut total = 0;

    for list in data
    {
        if is_valid(list)
        {
            total += 1;
        }
    }

    total
}

fn part2(data: &Vec<Vec<i32>>) -> i32
{
    let mut total = 0;

    for list in data
    {
        if is_valid(list)
        {
            total += 1;
        }
        else
        {
            for i in 0..list.len()
            {
                let mut new_list = list.clone();
                new_list.remove(i);
                if is_valid(&new_list)
                {
                    total += 1;
                    break;
                }
            }
        }
    }

    total
}

fn main()
{
    let file_path = "02_input.txt";
    match read_data(file_path)
    {
        Ok(result) =>
        {
            println!("{}", part1(&result));
            println!("{}", part2(&result));
        },
        Err(e) =>
        {
            eprintln!("Error: {}", e);
        }
    }
}